table_refresh.go

```diff
@@ -65,11 +65,58 @@ func (rt *RoutingTable) GenRandPeerID(targetCpl uint) (peer.ID, error) {
 
 	// Convert to a known peer ID.
 	key := keyPrefixMap[targetPrefix]
-	id := [34]byte{mh.SHA2_256, 32}
+	id := [Keysize + 2]byte{mh.SHA2_256, Keysize}
 	binary.BigEndian.PutUint32(id[2:], key)
 	return peer.ID(id[:]), nil
 }
 
+// GenRandomKey generates a random key matching a provided Common Prefix Length (Cpl) wrt. the local identity
+func (rt *RoutingTable) GenRandomKey(targetCpl uint) (ID, error) {
+	// targetCpl cannot be larger than the key size in bits
+	if targetCpl >= Keysize*8 {
+		return nil, fmt.Errorf("cannot generate peer ID for Cpl greater than key length")
+	}
+
+	// the generated key must match the targetCpl first bits of the local key,
+	// the following bit is the inverse of the local key's bit at position targetCpl+1
+	// and the remaining bits are randomly generated
+	//
+	// The returned ID takes the first targetCpl/8 bytes from the local key, the next byte is
+	// targetCpl%8 bits from the local key, 1 bit from the local key inverted and the remaining
+	// bits are random, and the last bytes are random
+
+	// generate random bytes
+	nRandBytes := Keysize - (targetCpl+1)/8
+	randBytes := make([]byte, nRandBytes)
+	_, err := rand.Read(randBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// byte and bit offset for the given cpl
+	byteOffset := targetCpl / 8
+	bitOffset := targetCpl % 8
+
+	// copy the local key to a new slice
+	randKey := make([]byte, len(rt.local))
+	copy(randKey, []byte(rt.local))
+
+	// compute the mask for the local key, the first targetCpl bits must be the same as the local key
+	// hence the mask is 1s for bits up to targetCpl and 0s for the rest
+	localMask := (^byte(0)) << (8 - bitOffset)
+	// compute the bit that is flipped in the local key (at position targetCpl+1)
+	bucketBit := (byte(0x80) >> bitOffset) & ^rt.local[targetCpl/8]
+
+	// first bitOffset bits are the same as the local key, the next bit is flipped and the remaining bits are random
+	randKey[byteOffset] = (localMask & randKey[byteOffset]) | bucketBit | (randBytes[0] & ((^localMask) >> 1))
+	// the remaining bytes are random
+	for i := uint(1); i < nRandBytes; i++ {
+		randKey[byteOffset+i] = randBytes[i]
+	}
+
+	return randKey, nil
+}
```

Comment:
```suggestion
// GenRandomKey generates a random key matching a provided Common Prefix Length (Cpl) wrt. the local identity
func (rt *RoutingTable) GenRandomKey(targetCpl uint) (ID, error) {
	if targetCpl%8 >= len(rt.local)-1 {
		return nil, fmt.Errorf("cannot generate peer ID for Cpl greater than key length")
	}

	output := make([]byte, len(rt.local))
	partialOffset := targetCpl / 8
	copy(output, rt.local[:partialOffset])
	_, err := rand.Read(output[partialOffset:])
	if err != nil {
		return nil, err
	}
	remainingBits := targetCpl % 8
	if remainingBits != 0 {
		origMask := ^uint8(0) << remainingBits
		randMask := ^origMask >> 1
		notMatchingBitOffset := remainingBits - 1
		notOrigMask := 1 << notMatchingBitOffset
		orig := rt.local[partialOffset]
		output[partialOffset] = orig&origMask | (orig & notOrigMask) ^ notOrigMask | output[partialOffset]&randMask // restore the high bits of the original key
	} else {
		output[partialOffset] ^= 0b1000_0000
	}

	return ID(output), nil
}
```
Comments are not good code.
Code documentation (such as header comments on function, global variables and fields) is good, but not inline comments (unless your thing is dark magic, or you link to relevent parts of the spec).

If your bitfidling code is is so confusing it require that many comments it's a problem with your code, unlike code which have fuzzers, unit tests and static analyzers we lack reliable tools to use transistors to check comments thus they often fell out of sync with the code as it evolve (maybe chatgpt will solve this :rofl: ?).

Reply:
The bit logic in your suggestion is incorrect, it should be:
```go
// GenRandomKey generates a random key matching a provided Common Prefix Length (Cpl) wrt. the local identity
func (rt *RoutingTable) GenRandomKey(targetCpl uint) (ID, error) {
	if int(targetCpl)+1 >= len(rt.local)*8 {
		return nil, fmt.Errorf("cannot generate peer ID for Cpl greater than key length")
	}
	output := make([]byte, len(rt.local))
	partialOffset := targetCpl / 8
	copy(output, rt.local[:partialOffset])
	_, err := rand.Read(output[partialOffset:])
	if err != nil {
		return nil, err
	}
	remainingBits := targetCpl % 8
	orig := rt.local[partialOffset]
	if remainingBits != 0 {
		origMask := ^uint8(0) << (8 - remainingBits)
		randMask := ^origMask >> 1
		notMatchingBitOffset := remainingBits + 1
		notOrigMask := byte(1) << (8 - notMatchingBitOffset)
		output[partialOffset] = orig&origMask | (orig & notOrigMask) ^ notOrigMask | output[partialOffset]&randMask // restore the high bits of the original key
	} else {
		output[partialOffset] = output[partialOffset]&0b0111_1111 | (rt.local[partialOffset] & 0b1000_0000) ^ 0b1000_0000
	}

	return ID(output), nil
}
```

I generally agree that self describing code is a good practice. However, in non-trivial operations like this one, inline comments help quickly grasping what the code is doing, and help spotting bugs.

Generally, it would be great if we can speed up folks getting familiar with the codebase and developers building on top of the stack with more human readable information. Most of the codebases I have worked with are really hard to read...

For instance, a developer willing to use this function in their code would need to dive in all bitwise operations to understand what this function is producing. And if they don't understand it correctly they won't use the function correctly. It can be easily avoided by describing with inline comments and/or longer function description.